### ISSUE TITLE
Reconfigure flyway migration params

### ISSUE THREAD

**Issue by alondhe** [OHDSI/Broadsea#85]
The current flyway migration params are hard-coded to establish a new install of WebAPI. Add options to allow for existing installs.

https://github.com/OHDSI/Broadsea/blob/main/ohdsi-webapi.yml#L33-L36

Tagging @chrisknoll

**Comment by alondhe** [comment]
We will remove these items.

**Comment by alondhe** [comment]
Ah okay. So we would need something in a profile -- default or in the atlasdb one -- that maintains this step of initializing the demo CDM as a source.

**Comment by chrisknoll** [comment]
@leeevans , couple things:

Flyway doesn't have anything to do with CDM databases, it only manages the WebAPI schema.   

there was something strange the way the 'baseline' flyway parameter was set.  Ajit said it works OK for new dbs, but existing ones have conflicts with duplicate sequences already being created, and it's probably somethign to do with the baseline function in Flyway. I looked at how basline works in Flyway, and it requires a special script to install to a specific version, which we never wrote for flyway. So, I think you should ignore the baseline functionality, and let flyway mgirations figure out which migrations to apply wheen it starts up.   If you pre-install a webAPI with part of the webapi database tables set up, with flyway, then if someone installs a new webapi on that instance, the webapi DB will maintain metadata inddicating which migrations have been done, and it should just work.

If you thought you could take a pre-loaded web api database tables, and tell flyway to ignore certain migrations using the 'baseline' functionality, I don't think it works that way  and instead you should just let flyway create all the tables.

**Comment by leeevans** [comment]
@chrisknoll thanks for the feedback.

broadsea-atlasdb contains a WebAPI schema and CDM database schemas - I should have mentioned that in my previous comment. 

The broadsea-atlasdb Dockerfile:

https://github.com/OHDSI/Broadsea-Atlasdb/blob/main/Dockerfile

rolls up the very early WebAPI schema migration scripts and executes them, just to the point where the source and source_daimon tables are created, so it can then populate those two tables in  the WebAPI schema before WebAPI starts up.

I set the flyway baseline parameter to indicate to flyway the last migration script that was executed in the broadsea-atlasdb migration script rollup, so when the ohdsi-webapi docker container starts up it runs just the subsequent migration scripts. I think I needed to set the baseline parameter because the flyway metadata table that tracks migration script execution isn't populated by the initial rollup migration sql execution as it would normally be set if flyway ran those migration scripts, so we have to tell flyway which was the last migration executed. Flyway updates that metadata table correctly when it runs the subsequent migration scripts.

None of this is needed for a new empty database or a database which has been created and upgraded normally by WebAPI.

**Comment by chrisknoll** [comment]
Hi, @leeevans ,
  I think I understand what you're doing about 'roling up very early WebAPI schema migration scripts and executes them', but I think I understand it to mean that you're executing this and not Flyway.   To put it another way, you're simulating a flyway migration.   Normally, I'd say 'don't do that' but I understand that you want the source daikon tables created so that you can pre-populate a cdm source.   

What I would do, instead, is make a special build that you get the records from webapi.schema_version that indicates which migrations were deployed, and insert the records into schema_version of the migration scripts you wanted to pre-execute.   Make that webapi databasse (with schema_version) your pre-packaged webapi database, such that when it starts up for real, it will identify the scripts have already been executed and only migrate stuff after.

I can work with you offline if you want to try to create a 'default WebAPI' db that has a partial migration completed.   Or we can look into how you set up the baseline migration and make it work with flyway, but even if you deploy the baseline, you still need to indicate in schema_vesrion that a baseline migration occurred.

**Comment by alondhe** [comment]
I'm still unclear here. If we want to have a ready-to-go CDM in Atlas, couldn't we simply have Broadsea deploy webapi to a blank schema, then run inserts into the source/source_daimon tables (and any other necessary inserts) to accomplish this? Then we wouldn't need to worry about migration scripts. This also allows us to provide a pattern where a user could want webapi with the ready-to-go CDM in a remote Postgres instance if need be.

**Comment by chrisknoll** [comment]
That's the trick:  how do you script it so that it will launch webAPI, _wait for the migrations to finish_, then perform the insert?  I think that was the challenge:   the source tables is early on in the 7-year migration history of WebAPI...an so I think Lee set it up to have the source table instantiated so that the source can be inserted, then the subsequent WebAPI startup does the rest of the migration.   

I don't know why we can't just have a fully-migrated webAPI db prepared in advance, but the 'baseline' described above seemed to put it into a 'minimum state' for getting a source table.  But I think it would be better to just start a webAPI in a Docker container, point it at an adjacent PG instance, let the WebAPI deploy the tables, insert the SOURCE records, and then snapshot the containers so that they're in a 'ready-to-go' state.

**Comment by alondhe** [comment]
Yeah, I'm thinking with Docker Compose offering some healthcheck options, we could have a health check on webapi starting successfully (maybe just being able to resolve the /info endpoint), then firing off inserts to source/source_daimon.

**Comment by leeevans** [comment]
I think of atlasdb as just one of the ‘batteries’ in the ‘batteries included’ OHDSI tool suite that Broadsea deploys. It’s optional - folks can replace it with their own postgres webapi database and they should replace it for production environments. Therefore, I think we should isolate all the config and supporting logic for atlasdb in the atlasdb profile and keep the Docker compose file as simple as possible - as we do with Broadsea 3.0 today.

**Comment by leeevans** [comment]
@chrisknoll I wanted to be completely transparent about how the atlasdb database is built and populated, so I opted for including and executing the sql files in the Dockerfile build process instead of providing just a snapshot image of the container. Anyone can review all the steps that are executed and build it themselves (and customize it) and I thought that would be a benefit for sites that have more stringent application security needs (e.g. security audit required).

**Comment by alondhe** [comment]
We can keep atlasdb as-is, but the question is on the WebAPI container, that's where the flyway stuff is being defined.

We could expose the flyway env vars to the user, but I find this to be challenging for user experience, as it wouldn't be very clear on what to set. 

We could offer another profile choice, but we already have webapi-from-image and webapi-from-git. Adding another dimension here could make the webapi profile options unwieldy.

I think instead that we could add a container service "broadsea-atlasdb-init" that waits for the /info endpoint of WebAPI to be available, then runs a few SQL scripts included in Broadsea to add the sources.

**Comment by chrisknoll** [comment]
If we want to get on a call to discuss this, I'd be happy to.  I'm not quite sure the need for having a post-launch step to insert records (which happens only once) vs. just pre-baking the 'included postgres DB' with the webAPI migrations already deployed + necessary source records that points to the sample CDM.
